File: 2025/day5/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

type Range struct {
	start int
	end   int
}

func main() {
	file, _ := os.Open("inputs.txt")
	defer file.Close()

	scanner := bufio.NewScanner(file)

	var ranges []Range
	var numbers []int

	reading_ranges := true

	for scanner.Scan() {
		line := scanner.Text()

		if line == "" {
			reading_ranges = false
			continue
		}

		if reading_ranges {
			var s, e int
			fmt.Sscanf(line, "%d-%d", &s, &e)
			ranges = append(ranges, Range{s, e})
		} else {
			var n int
			fmt.Sscanf(line, "%d", &n)
			numbers = append(numbers, n)
		}
	}

	// Part 1
	count := 0
	for _, n := range numbers {
		for _, r := range ranges {
			if n >= r.start && n <= r.end {
				count++
				break
			}
		}
	}

	fmt.Println("Part1:", count)

	// PART 2
	merged := mergeRanges(ranges)

	total := 0
	for _, r := range merged {
		total += r.end - r.start + 1
	}

	fmt.Println("Part2:", total)
}

func mergeRanges(ranges []Range) []Range {
	sort.Slice(ranges, func(i, j int) bool {
		return ranges[i].start < ranges[j].start
	})

	merged := []Range{}

	for _, r := range ranges {
		if len(merged) == 0 {
			merged = append(merged, r)
		} else {
			last := &merged[len(merged)-1]

			if r.start <= last.end {
				if r.end > last.end {
					last.end = r.end
				}
			} else {
				merged = append(merged, r)
			}
		}
	}
	return merged
}

File: 2025/day5/main.py
ranges = []
numbers = []

reading_ranges = True

# get the data
with open("inputs.txt") as f:
    for line in f:
        line = line.strip()

        if not line:
            reading_ranges = False
            continue

        if reading_ranges:
            s, e = map(int, line.split("-"))
            ranges.append((s, e))
        else:
            numbers.append(int(line))

# Part 1
count = 0

for n in numbers:
    for s, e in ranges:
        if s <= n <= e:
            count += 1
            break
print("Part 1: ", count)


# Part 2
ranges.sort()

merged = []
for s, e in ranges:
    if not merged or s > merged[-1][1]:
        merged.append([s, e])
    else:
        merged[-1][1] = max(merged[-1][1], e)

total = sum(e - s + 1 for s, e in merged)
print("Part 2: ", total)
